jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java

```diff
@@ -34,42 +36,52 @@
 
   private final AuthParameters auth;
 
-  @Nullable private String image;
-  private Set<String> tags = Collections.emptySet();
+  private Property<String> image;
+  private SetProperty<String> tags;
   @Nullable private String credHelper;
 
   @Inject
   public TargetImageParameters(ObjectFactory objectFactory) {
     auth = objectFactory.newInstance(AuthParameters.class, "to.auth");
+    image = objectFactory.property(String.class);
+    tags = objectFactory.setProperty(String.class).empty();
   }
 
   @Input
   @Nullable
-  @Optional
   public String getImage() {
     if (System.getProperty(PropertyNames.TO_IMAGE) != null) {
       return System.getProperty(PropertyNames.TO_IMAGE);
     }
-    return image;
+    return image.getOrNull();
   }
 
   public void setImage(String image) {
-    this.image = image;
+    this.image.set(image);
+  }
+
+  public void setImage(Provider<String> image) {
+    if (image != null) {
+      this.image.set(image);
+    }
   }
 
   @Input
-  @Optional
   public Set<String> getTags() {
     if (System.getProperty(PropertyNames.TO_TAGS) != null) {
       return ImmutableSet.copyOf(
           ConfigurationPropertyValidator.parseListProperty(
               System.getProperty(PropertyNames.TO_TAGS)));
     }
-    return tags;
+    return tags.get();
   }
 
   public void setTags(Set<String> tags) {
-    this.tags = tags;
+    this.tags.set(tags);
+  }
+
+  public void setTags(Provider<Set<String>> tags) {
+    this.tags.set(tags);
```

Comment:
In light of https://github.com/GoogleContainerTools/jib/pull/2739/files#r485008215, I wonder if we need a null check. Can you run a manual test and see which of the two `setTags()` gets called?

Reply:
That is a good idea. So unlike the case with the image, setting `jib.to.tags` to null is resulting in GroovyRuntimeException with `Ambiguous method overloading...Cannot resolve which method to invoke for [null] due to overlapping prototypes between` where one `setTags` is not preferred over the other. The issue occurs when _only_ `setTags(Provider<Set<String>>)` is present, which is when setting `jib.to.tags=null` results in the `Cannot set the value of property using null provider` message.

Reply:
Passing in a set or provider<set<string>> that _contains_ `null` seems to resulting in NPE so I think we need to handle that?

Reply:
Hmm... I think we should be fine as long as nothing breaks when `jib.to.tags` is missing (not configured). People shouldn't explicitly set it to `null` anyway. That said, I now wonder the test failure you mentioned in https://github.com/GoogleContainerTools/jib/pull/2739/files#r485008215 is when you explicitly set the property to `null` rather than when it's missing. If former, I don't think we need the null check there either. 

> Passing in a set or provider<set> that contains null seems to resulting in NPE so I think we need to handle that?

Let's file an issue for that. It basically sounds like an orthogonal issue we don't want to mix in here.

Reply:
Right that's a good point. I filed an issue for it in #2760